### PR TITLE
Add empty implementations for abstract functions

### PR DESCRIPTION
**Pull request by seanvk**
The issue is that traditional cmrt applications (not using dynamic loading, but link at build time) failed to build with cmrt library due to link error such as:
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:146: undefined reference to `CmDevice::CreateSurface2D(un
signed int, unsigned int, _VA_CM_FORMAT, CmSurface2D*&)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:153: undefined reference to`CmSurface2D::WriteSurface(un
signed char const_, CmEvent_, unsigned long)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:160: undefined reference to `CmDevice::CreateSurface2D(un
signed int, unsigned int, _VA_CM_FORMAT, CmSurface2D*&)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:167: undefined reference to`CmDevice::CreateSurface2D(un
signed int, unsigned int, _VA_CM_FORMAT, CmSurface2D_&)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:205: undefined reference to `CmDevice::LoadProgram(void_,
unsigned int, CmProgram_&, char const_)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:213: undefined reference to `CmDevice::CreateKernel(CmPro
gram*, char const*, CmKernel*&, char const*)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:223: undefined reference to`CmKernel::SetThreadCount(uns
igned int)'
/work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:231: undefined reference to `CmSurface2D::GetIndex(Surfac
eIndex*&)'

Signed-off-by: Sean V Kelley seanvk@posteo.de

### CODE REVIEW

**Comment by guoyejun** [comment]
Sorry, my fault.

I tried cmrt unit tests which is a package that already includes the header files, so my experiments are based on the old header files but link against the new library …

Everything is OK

**Comment by jayesht** [comment]
Validation this pull request on Braswell Ubuntu platform for VP9 decode tests. 95/108 test pass. Rest 13 failures are already known failures. Good to merge.

**Comment by gbeauchesne** [comment]
Hi,

2015-10-09 21:54 GMT+02:00 Sean V. Kelley notifications@github.com:

> The issue is that traditional cmrt applications (not using dynamic
> loading, but link at build time) failed to build with cmrt library due to
> link error such as:
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:146:
> undefined reference to CmDevice::CreateSurface2D(un
> signed int, unsigned int, _VA_CM_FORMAT, CmSurface2D_&)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:153:
> undefined reference toCmSurface2D::WriteSurface(un
> signed char const_, CmEvent_, unsigned long)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:160:
> undefined reference to CmDevice::CreateSurface2D(un
> signed int, unsigned int, _VA_CM_FORMAT, CmSurface2D_&)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:167:
> undefined reference toCmDevice::CreateSurface2D(un
> signed int, unsigned int, _VA_CM_FORMAT, CmSurface2D
> _&)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:205:
> undefined reference to `CmDevice::LoadProgram(void_,
> unsigned int, CmProgram_&, char const_)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:213:
> undefined reference to CmDevice::CreateKernel(CmPro
> gram_, char const_, CmKernel_&, char const_)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:223:
> undefined reference toCmKernel::SetThreadCount(uns
> igned int)'
> /work/cmrt/otc_version/tests_from_itflex/Examples/sharpen/sharpen/sharpen.cpp:231:
> undefined reference to `CmSurface2D::GetIndex(Surfac
> eIndex*&)'
> 
> ## Signed-off-by: Sean V Kelley seanvk@posteo.de
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/01org/cmrt/pull/25
> Commit Summary
> - Add empty implementations for abstract functions
> 
> File Changes
> - _M_ src/cm_device_base.h
>   https://github.com/01org/cmrt/pull/25/files#diff-0 (50)
> 
> Patch Links:
> - https://github.com/01org/cmrt/pull/25.patch
> - https://github.com/01org/cmrt/pull/25.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/01org/cmrt/pull/25.

This doesn't look right. If the base methods were pure, there was a reason
for it. Presumably to make sure that actual users don't instantiate through
CmDevice but probably CmDevice_RT?  This means, fixes are rather needed in
the application part. Otherwise, you now allow instantiation from CmDevice,
which provides nothing. This is a call to errors.

In other words, CmDevice is no longer an Abstract Base Class. This is an
API change that further needs to be documented, if this change was really

## needed. Which is doubtful.

Gwenole Beauchesne
Intel Corporation SAS / 2 rue de Paris, 92196 Meudon Cedex, France
Registration Number (RCS): Nanterre B 302 456 199

**Comment by guoyejun** [comment]
yes, you are right. I made a mistake, and explained that "Everything is OK" in the comments (see above).
